commit e99bf7484e4c7d8d154a20a12327963be6a68cfc
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Sat Dec 31 18:19:03 2022 -0600

    fix all build warnings

diff --git a/Makefile b/Makefile
index e551806..ff68852 100644
--- a/Makefile
+++ b/Makefile
@@ -11,9 +11,8 @@ ARMOBJCOPY		=	$(ARMTOOLCHAIN)objcopy
 LTO				=	#-flto
 ARMLTO			=	-flto
 ARMTYPE			=	-marm		#shoudl be -mthumb or -marm
-COMMON			=	-DPNGLE_NO_GAMMA_CORRECTION -DPNGLE_SKIP_CRC
-M68KCOMMON		=	$(COMMON) -Wno-multichar -funsafe-math-optimizations -Os -m68000 -mno-align-int -mpcrel -fpic -fshort-enums -mshort
-ARMCOMMON		=	$(COMMON) -Ofast -march=armv4t $(ARMTYPE) -mno-unaligned-access -ffixed-r9 -ffixed-r10 -ffixed-r11 -fomit-frame-pointer -D__ARM__ -ffreestanding -fpic -mthumb-interwork
+M68KCOMMON		=	$(COMMON) -Wno-multichar -funsafe-math-optimizations -Os -m68000 -mno-align-int -mpcrel -fpic -fshort-enums -mshort -fvisibility=hidden -Wno-attributes
+ARMCOMMON		=	$(COMMON) -Ofast -march=armv4t $(ARMTYPE) -mno-unaligned-access -ffixed-r9 -ffixed-r10 -ffixed-r11 -fomit-frame-pointer -D__ARM__ -ffreestanding -fpic -mthumb-interwork -Wno-attributes
 WARN			=	-Wsign-compare -Wextra -Wall -Wno-unused-parameter -Wno-old-style-declaration -Wno-unused-function -Wno-unused-variable -Wno-error=cpp -Wno-error=switch
 LKR				=	Src/68k.lkr
 ARMLKR			=	src/arm.lkr
diff --git a/Src/PkmnMain.c b/Src/PkmnMain.c
index 9345976..1e7cd4b 100644
--- a/Src/PkmnMain.c
+++ b/Src/PkmnMain.c
@@ -106,7 +106,7 @@ void SetDescriptionField(UInt16 selectedPkmnId)
 	char *text = pokeDescrGet(selectedPkmnId);
 	
 	if (!text)
-		text = emptyString;
+		text = (char*)emptyString;
 
 	FldSetTextPtr(fld, text);
 	FldRecalculateField(fld, true);
@@ -117,9 +117,9 @@ void FreeDescriptionField()
 	FieldType *fld = GetObjectPtr(PkmnMainDescField);
 	void *ptr = FldGetTextPtr(fld);
 	
-	FldSetTextPtr(fld, emptyString);
+	FldSetTextPtr(fld, (char*)emptyString);
 	
-	if (ptr && ptr != emptyString)
+	if (ptr && ptr != (char*)emptyString)
 		MemPtrFree(ptr);
 }
 
diff --git a/Src/osPatches.68k.o b/Src/osPatches.68k.o
new file mode 100644
index 0000000..ea58214
Binary files /dev/null and b/Src/osPatches.68k.o differ
