commit f97a18000667f9b44f5f70bd4c6a497f6d8ef8dd
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Mon Dec 26 23:37:30 2022 -0600

    remove WinPopDrawState

diff --git a/Src/PkmnType.c b/Src/PkmnType.c
index 81e3cf6..e35ea3a 100644
--- a/Src/PkmnType.c
+++ b/Src/PkmnType.c
@@ -127,7 +127,6 @@ static void DrawEffectiveness(UInt16 selectedPkmnID, UInt8 x, UInt8 y, UInt8 typ
     
 		MemPtrFree(str);
 	}
-    WinPopDrawState();
 }
 
 static void DrawTypeIcons(UInt16 selectedPkmnID)
