commit 612afa6fd0cc67881d679f515deed87b5ff383c5
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Tue Dec 27 00:44:24 2022 -0600

    first part of PalmOs 1.0 compat

diff --git a/Rsc/Palmkedex_Rsc.rcp b/Rsc/Palmkedex_Rsc.rcp
index a5574e4..799f752 100644
--- a/Rsc/Palmkedex_Rsc.rcp
+++ b/Rsc/Palmkedex_Rsc.rcp
@@ -7,6 +7,29 @@ DATA "code" ID 1 "code0001.68k.bin"
 DATA "armc" ID 1 "armc0001.arm.bin"
 GENERATEHEADER "Rsc/Palmkedex_Rsc.h"
 
+// these next two resources are usually optional, except in PalmOS 1.0, where they are not...
+HEX "code" ID 0
+	0x00 0x00 0x00 0x04
+	0x00 0x00 0x00 0x00
+
+HEX "data" ID 0
+	0x00 0x00 0x00 0x00		//ignored if data res ends after init parts
+	
+	//init data 1/3
+	0x00 0x00 0x00 0x00		//offset to start of "initing"
+	0x00					//END of init data
+	
+	//init data 2/3
+	0x00 0x00 0x00 0x00		//offset to start of "initing"
+	0x00					//END of init data
+	
+	//init data 3/3
+	0x00 0x00 0x00 0x00		//offset to start of "initing"
+	0x00					//END of init data
+	
+	
+	
+
 // Use this for CodeWarrior
 //GENERATEHEADER "Rsc/Palmkedex_Rsc.h"
 
diff --git a/Src/Palmkedex.c b/Src/Palmkedex.c
index 2822ca1..46e1a5b 100644
--- a/Src/Palmkedex.c
+++ b/Src/Palmkedex.c
@@ -33,7 +33,7 @@
  *********************************************************************/
 
 /* Define the minimum OS version we support */
-#define ourMinVersion    sysMakeROMVersion(2,0,0,sysROMStageDevelopment,0)
+#define ourMinVersion    sysMakeROMVersion(1,0,0,sysROMStageDevelopment,0)
 #define kPalmOS20Version sysMakeROMVersion(2,0,0,sysROMStageDevelopment,0)
 
 /*********************************************************************
@@ -59,9 +59,17 @@
 void * GetObjectPtr(UInt16 objectID)
 {
 	FormType * frmP;
+	UInt16 idx;
 
 	frmP = FrmGetActiveForm();
-	return FrmGetObjectPtr(frmP, FrmGetObjectIndex(frmP, objectID));
+	if (!frmP)
+		return NULL;
+	
+	idx = FrmGetObjectIndex(frmP, objectID);
+	if (idx == frmInvalidObjectId)
+		return NULL;
+	
+	return FrmGetObjectPtr(frmP, idx);
 }
 
 
@@ -336,10 +344,14 @@ static Err RomVersionCompatible(UInt32 requiredVersion, UInt16 launchFlags)
 
 static Err loadSonyHrLib(UInt16 *hrLibRefP)
 {
-	UInt32 val320 = 320;
+	UInt32 val320 = 320, romVersion;
 	UInt16 hrLibRef;
 	Err e;
-			
+	
+	//Library loading is broken on OS1, and no sony devices shipped with less than 3, so give up early
+	if (errNone != FtrGet(sysFtrCreator, sysFtrNumROMVersion, &romVersion) || romVersion < sysMakeROMVersion(3,0,0,sysROMStageDevelopment,0))
+		return sysErrLibNotFound;
+	
 	e = SysLibFind(sonySysLibNameHR, &hrLibRef);
 	if (e == sysErrLibNotFound)
 		e = SysLibLoad(sysFileTLibrary, sonySysFileCHRLib, &hrLibRef);
