commit e567228fcc72c76f42e020b2ff6ec65e88dacc44
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Wed Dec 28 00:11:30 2022 -0600

    remove more remnants of the scrollbar code

diff --git a/Src/PkmnMain.c b/Src/PkmnMain.c
index d3fd662..259dec4 100644
--- a/Src/PkmnMain.c
+++ b/Src/PkmnMain.c
@@ -217,19 +217,6 @@ static void PkmnDescriptionScroll(WinDirectionType direction)
 	}
 }
 
-static void PkmnDescriptionSimpleScroll(Int16 linesToScroll)
-{
-	FieldPtr fld;
-
-	fld = GetObjectPtr(PkmnMainDescField);
-
-	if (linesToScroll < 0)
-		FldScrollField(fld, -linesToScroll, winUp);
-
-	else if (linesToScroll > 0)
-		FldScrollField(fld, linesToScroll, winDown);
-}
-
 static void unregisterCurrentPng()
 {
 	struct DrawState *ds;
@@ -320,11 +307,6 @@ Boolean PkmnMainFormHandleEvent(EventType *eventP)
 		}
 		break;
 
-	case sclRepeatEvent:
-		PkmnDescriptionSimpleScroll(eventP->data.sclRepeat.newValue -
-									eventP->data.sclRepeat.value);
-		break;
-
 	case popSelectEvent:
 		if (eventP->data.popSelect.selection == 1)
 		{
