commit 5604dbd53f8e61493425d1e6b5661680a3b6e5d8
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Mon Dec 26 23:36:25 2022 -0600

    remove more WinPaintChars

diff --git a/Src/Main.c b/Src/Main.c
index 025154f..0a6b533 100644
--- a/Src/Main.c
+++ b/Src/Main.c
@@ -57,7 +57,7 @@ static void PokemonListDraw(Int16 itemNum, RectangleType *bounds, Char **unused)
 	FntSetFont(boldFont);
 	WinDrawChars(sharedVars->pkmnLstNumStr, 4, bounds->topLeft.x, bounds->topLeft.y);
 	FntSetFont(stdFont);
-	WinPaintChars(sharedVars->pkmnLstNameStr, MAX_PKMN_NAME_LEN, bounds->topLeft.x + 32, bounds->topLeft.y);
+	WinDrawChars(sharedVars->pkmnLstNameStr, MAX_PKMN_NAME_LEN, bounds->topLeft.x + 32, bounds->topLeft.y);
 }
 
 static void ParseSearchString(Char *searchStr, Char charInserted)
