commit b866ca9628f3161c34c9fd52766bb2c541d90e9e
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Wed Dec 28 00:56:54 2022 -0600

    small ui fix in type screen

diff --git a/Rsc/Palmkedex_Rsc.rcp b/Rsc/Palmkedex_Rsc.rcp
index bf6f93c..1009245 100644
--- a/Rsc/Palmkedex_Rsc.rcp
+++ b/Rsc/Palmkedex_Rsc.rcp
@@ -148,9 +148,11 @@ NOFRAME
 BEGIN
     TITLE "Loading..."
 	
-	POPUPTRIGGER "Type" PkmnTypePopUpTrigger AT (119 1 AUTO AUTO)
-	LIST "General" "Type" PkmnTypePopUpList AT (PREVLEFT PREVBOTTOM+1 PREVWIDTH 35) VISIBLEITEMS 2 NONUSABLE 
-	POPUPLIST PkmnTypePopUpTrigger PkmnTypePopUpList
+	
+	POPUPTRIGGER "Type" ID PkmnTypePopUpTrigger  AT (121 1 39 12)  LEFTANCHOR
+	LIST  "Info" "Type" ID PkmnTypePopUpList  AT (135 1 25 24) VISIBLEITEMS 2 NONUSABLE
+	POPUPLIST ID PkmnTypePopUpTrigger PkmnTypePopUpList
+	
 END
 
 //████████ ██    ██ ██████  ███████ ███████     ██████  ███    ███ ██████  ███████ 
