commit c6b1cdb0ad0c46ac3d7e1ac6362a42a5fd28684d
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Tue Dec 27 00:20:42 2022 -0600

    fix bug on exist if no sprites on device

diff --git a/Src/PkmnMain.c b/Src/PkmnMain.c
index c28c143..b4c1e38 100644
--- a/Src/PkmnMain.c
+++ b/Src/PkmnMain.c
@@ -260,8 +260,7 @@ static void unregisterCurrentPng()
 {
 	struct DrawState *ds;
 	
-	FtrGet(appFileCreator, 0, (UInt32*)&ds);
-	if (ds)
+	if (FtrGet(appFileCreator, 0, (UInt32*)&ds) == errNone && ds)
 	{
 		imgDrawStateFree(ds);
 		FtrUnregister(appFileCreator, 0);
