commit f191d6a051cb62235ce332dd765d35daea5cdbbc
Author: Dmitry Grinberg <dmitrygr@gmail.com>
Date:   Thu Dec 22 22:41:49 2022 -0600

    fix helpers for better posix compat

diff --git a/Src/helpers.c b/Src/helpers.c
index e944c3f..002d360 100644
--- a/Src/helpers.c
+++ b/Src/helpers.c
@@ -21,8 +21,21 @@ void* malloc(UInt32 sz) { return MemPtrNew(sz); }
 void free(void *p) { if(p) MemPtrFree(p); }
 
 void* realloc(void *p, UInt32 sz) { 
-    MemPtrResize(p, sz); 
-    return p;
+    
+    void *newPtr;
+    
+    if (errNone == MemPtrResize(p, sz))
+    	return p;
+    
+    //fail means we ARE upsizing!
+    
+    newPtr = MemPtrNew(sz);
+    if (!newPtr)
+    	return NULL;
+    
+    MemMove(newPtr, p, MemPtrSize(p));
+    MemPtrFree(p);
+    return newPtr;
 }
 
 void* calloc(UInt32 qtty, UInt32 sz) { 
@@ -32,9 +45,13 @@ void* calloc(UInt32 qtty, UInt32 sz) {
 }
 
 void* memset (void *p, int c, UInt32 l) {
-    return MemSet(p, l, c);
+    MemSet(p, l, c);
+    
+    return p;
 }
 
 void* memcpy (void *d, const void *s, UInt32 l) {
-    return MemMove(d, s, l);
+    MemMove(d, s, l);
+    
+    return d;
 }
